src/Analyser/MutatingScope.php

```diff
@@ -4489,7 +4489,7 @@ public function addTypeToExpression(Expr $expr, Type $type): self
 
 		if ($originalExprType->equals($nativeType)) {
 			$newType = TypeCombinator::intersect($type, $originalExprType);
-			if ($newType->isConstantScalarValue()->yes() && $newType->equals($originalExprType)) {
+			if ($newType->isObject()->no() && $newType->equals($originalExprType)) {
```

Comment:
Will look into whether we can get rid of the isObject() check here today

Reply:
ok, I looked into it. I think we cannot drop the `$newType->isObject()->no()` check here, because Objects will be considered equal by `ObjectType->equals()` even though they have slight differences which are important when narrowing, e.g.

`new Foo()` in local-scope vs. `Foo` typed variable

Reply:
most conservative move would be to leave it `$newType->isConstantScalarValue()->yes()` until we get a slow case which benefits from the change and only fix `MixedType->equals(ErrorType)` for now